src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1558,3 +2324,391 @@ MultiRouterPlannableQuery(Query *query, MultiExecutorType taskExecutorType,
+
+		if (targetVarCount == 1)
+		{
+			Var *oldInsertVar = (Var *) linitial(targetVarList);
+			TargetEntry *oldSubqueryTle = list_nth(subquery->targetList,
+												   oldInsertVar->varattno - 1);
+
+			newSubqueryTargetEntry = copyObject(oldSubqueryTle);
+
+			newSubqueryTargetEntry->resno = resno;
+			newSubqueryTargetlist = lappend(newSubqueryTargetlist,
+											newSubqueryTargetEntry);
+
+			updatedSubqueryEntryCount++;
+		}
+		else
+		{
+			newSubqueryTargetEntry = makeTargetEntry(oldInsertTargetEntry->expr,
+													 resno,
+													 oldInsertTargetEntry->resname,
+													 oldInsertTargetEntry->resjunk);
+			newSubqueryTargetlist = lappend(newSubqueryTargetlist,
+											newSubqueryTargetEntry);
+		}
+
+		newInsertVar = makeVar(insertTableId, originalAttrNo,
+							   exprType((Node *) newSubqueryTargetEntry->expr),
+							   exprTypmod((Node *) newSubqueryTargetEntry->expr),
+							   exprCollation((Node *) newSubqueryTargetEntry->expr),
+							   0);
+		newInsertTargetEntry = makeTargetEntry((Expr *) newInsertVar, originalAttrNo,
+											   oldInsertTargetEntry->resname,
+											   oldInsertTargetEntry->resjunk);
+
+		newInsertTargetlist = lappend(newInsertTargetlist, newInsertTargetEntry);
+		resno++;
+	}
+
+	/*
+	 * if there are any remaining target list entries (i.e., GROUP BY column not on the
+	 * target list of subquery), update the remaining resnos.
+	 */
+	subqueryTargetLength = list_length(subquery->targetList);
+	if (subqueryTargetLength != updatedSubqueryEntryCount)
+	{
+		int targetEntryIndex = updatedSubqueryEntryCount;
+
+		for (; targetEntryIndex < subqueryTargetLength; ++targetEntryIndex)
+		{
+			TargetEntry *oldSubqueryTle = list_nth(subquery->targetList,
+												   targetEntryIndex);
+			TargetEntry *newSubqueryTargetEntry = copyObject(oldSubqueryTle);
+
+			Assert(newSubqueryTargetEntry->resjunk == true);
+
+			newSubqueryTargetEntry->resno = resno;
+			newSubqueryTargetlist = lappend(newSubqueryTargetlist,
+											newSubqueryTargetEntry);
+
+			resno++;
+		}
+	}
+
+	originalQuery->targetList = newInsertTargetlist;
+	subquery->targetList = newSubqueryTargetlist;
+
+	return NULL;
+}
+
+
+/*
+ * InsertSelectQuery returns true when the input query
+ * is INSERT INTO ... SELECT kind of query.
+ *
+ * Note that the input query should be the original parsetree of
+ * the query (i.e., not passed trough the standard planner).
+ *
+ * This function is inspired from getInsertSelectQuery() on
+ * rewrite/rewriteManip.c.
+ */
+bool
+InsertSelectQuery(Query *query)
+{
+	CmdType commandType = query->commandType;
+	List *fromList = NULL;
+	RangeTblRef *rangeTableReference = NULL;
+	RangeTblEntry *subqueryRte = NULL;
+	Query *selectQuery = NULL;
+
+	if (commandType != CMD_INSERT)
+	{
+		return false;
+	}
+
+	if (query->jointree == NULL || !IsA(query->jointree, FromExpr))
+	{
+		return false;
+	}
+
+	fromList = query->jointree->fromlist;
+	if (list_length(fromList) != 1)
```

Comment:
Hmm, once we have `INSERT INTO uq_two_columns (partition_col, other_col) VALUES (1,1), (1,1);`
`list_length(fromList)` becomes 1 and  rteKind becomes `RTE_VALUES`. So, I believe we should keep that `if`s, instead of Assert?